### ISSUE TITLE
Add a `swiftlint config` subcommand

### ISSUE THREAD

**Issue by AliSoftware** [realm/SwiftLint#532]
I think adding a `swiftlint config` subcommand would be great, here is what it would be able to do:
1. Get help on the syntax of the configuration file, using `swiftlint config --help` to print the various keys allowed in the `.swiftlint.yml` and a reminder of what they mean
2. Initialize a template `.swiftlint.yml` file using `swiftlint config --init`. That would replace the `init` subcommand proposed by #518, basically doing what is proposed there but unifying it under the "config" task.
3. One might also consider `swiftlint config --example` to implement what is suggested in #517

---

Note that having this dedicated subcommand `swiftlint config` will also allow us to later add other actions related to the config file, so it makes sense to have that subcommand to group them instead of having a separated `swiftlint init` for example. See #533 for ideas of future subcommands.

**Comment by jpsim** [comment]
It'd be interesting to have the ability to "guess" which rules/parameters apply to a given project.

For example, if we see that there are many lines that are _almost_ 160 but that never exceed it, we know there's a high probability that's an existing stylistic choice.

Tabs vs space. Explicit vs implicit ACLs. There's quite a bit that could be "guessed" from a project.

**Comment by marcelofabri** [comment]
Another possibility is creating a configuration file with all needed configurations so the current project lints. IIRC rubocop does that. That might be useful for adding SwiftLint to an existing project as one could re-enable the rules slowly.

**Comment by jpsim** [comment]
Rubocop's inferred configuration feature is what prompted me to say this. 😊

**Comment by AliSoftware** [comment]
I love the idea. But maybe that could be implemented in multiple steps, to avoid postponing it too much?
Like start by implementing `swiftlint config init` at first, then add `swiftlint config help` then.
And add `swiftlint config guess` only later, as that last one seems more tricky to me (having to gather stats like max value for numeric rules like `line_length`, but count of violations vs non-violations for binary rules like `trailing_comma`…  and numeric rules might actually be more complicated that just 'max' if we want to guess both the warning and the error levels (at least for rules that can configure both of them)

**Comment by marcelofabri** [comment]
Totally agree 👍

**Comment by jpsim** [comment]
See `rubocop --auto-gen-config`: https://github.com/bbatsov/rubocop/blob/master/manual/configuration.md#automatically-generated-configuration

> If you have a code base with an overwhelming amount of offenses, it can
> be a good idea to use `rubocop --auto-gen-config` and add an
> `inherit_from: .rubocop_todo.yml` in your `.rubocop.yml`. The generated
> file `.rubocop_todo.yml` contains configuration to disable cops that
> currently detect an offense in the code by excluding the offending
> files, or disabling the cop altogether once a file count limit has been
> reached.
>
> By adding the option `--exclude-limit COUNT`, e.g., `rubocop
> --auto-gen-config --exclude-limit 5`, you can change how many files are
> excluded before the cop is entirely disabled. The default COUNT is 15.
>
> Then you can start removing the entries in the generated
> `.rubocop_todo.yml` file one by one as you work through all the offenses
> in the code.

**Comment by possen** [comment]
Question, has this been implemented somewhere? If not what is the status?

If not, I was wondering why it could not just be a parameter on the existing `rules` subcommand `--output-yml` or something, you could generate all disabled and all enabled by passing the existing parameters. Maybe extensions such as `--opt-in`, etc.  For example, if there was no config file, it would output all the defaults YML. 

`    swiftlint rules --output-yml --opt-in=true`

would generate output to the stdout if you wanted to write it you would do: 

`    swiftlint rules --output-yml --opt-in=true > .swiftlint.yml ` 

If there were a config file then it would continue to work as it does now which is use the values from the existing YML file. 
 
options would be:

| Option  | Description |
| ----------------------------------------- | -----------------------------------------|
| --correctable=[true/false] |   outputs only options that match |
| --kind=[lint/style/idiomatic/performance] |  outputs only options that match |
| --output-yml |  outputs in .yml format |
| --opt-in=[true/false] | outputs only options that match |
| --enabled |  outputs only ones enabled in your config |
| --disabled |  outputs only ones disabled in your config |
| --type=[warning/error] |  outputs only those that match |
| --config  path | specify the path to a swiftlint.yml |  

The configurable parameters would be output in YML, that could used for your config and simply edit it if you want to make changes. 

Is this idea worth pursuing?

**Comment by jpsim** [comment]
> Question, has this been implemented somewhere? If not what is the status?

No it hasn't been implemented. No one has started on this as far as I can tell. I'd rather `config` be a top-level command rather than flags on `rules`. My preference is for something closer to what rubocop has with `--auto-gen-config`.

**Comment by possen** [comment]
I like the idea of `--auto-gen-config` in that it intelligently turns off features and you set the limits for when you turn off features. 

Just exploring options... it seems like that could be accommodated with the above suggestion as well. I only combined it with the existing command because they seemed parallel and conceptually very similar, and that is where I first looked when I thought about generating the .yml file, maybe just me though. I then spent time writing a reg-ex to parse the output of that file so I could create a .yml file from it. 

I was concerned about duplication of code as separate commands and conceptually having to learn two things that could work similarly. The other nice benefit is that it would make the existing table output command more flexible and more powerful by having various "query" parameters. Not trying to be pushy here, just want to explain my thinking in suggesting the idea. 

So, if the command were a top level command and had the features I list above plus the auto flag, would that be desirable? Or does adding that feature to `rules` make more sense now that, I hope, explained it more. Or is my suggesting of the output to a file the part you thought was problematic? 

thanks.
